### ISSUE TITLE
gcc + __attribute__((deprecated))

### ISSUE THREAD

**Issue by karel-m** [libtom/libtommath#322]
Building with gcc-3.4.5 fails with error:
```
tommath.h:234: error: wrong number of arguments specified for `deprecated' attribute
```

It looks like gcc 3.x does support:
```
__attribute__((deprecated))
```
but not:
```
__attribute__((deprecated("some message")))
```

**Comment by karel-m** [comment]
IMO `__attribute__((deprecated("some message")))` is supported since gcc-4.5.0 see:
* https://gcc.gnu.org/onlinedocs/gcc-4.5.0/gcc/Function-Attributes.html (supported)
* https://gcc.gnu.org/onlinedocs/gcc-4.4.7/gcc/Function-Attributes.html (unsupported)

So perhaps:
```diff
- #if defined(__GNUC__) && (__GNUC__ * 100 + __GNUC_MINOR__ >= 301)
+ #if defined(__GNUC__) && (__GNUC__ * 100 + __GNUC_MINOR__ >= 405)
```

**Comment by karel-m** [comment]
Or more complicated:
```patch
- #if defined(__GNUC__) && (__GNUC__ * 100 + __GNUC_MINOR__ >= 301)
+ #if defined(__GNUC__) && (__GNUC__ * 100 + __GNUC_MINOR__ >= 405)
  #  define MP_DEPRECATED(x) __attribute__((deprecated("replaced by " #x)))
  #  define PRIVATE_MP_DEPRECATED_PRAGMA(s) _Pragma(#s)
  #  define MP_DEPRECATED_PRAGMA(s) PRIVATE_MP_DEPRECATED_PRAGMA(GCC warning s)
+ #elif defined(__GNUC__) && (__GNUC__ * 100 + __GNUC_MINOR__ >= 301)
+ #  define MP_DEPRECATED(x) __attribute__((deprecated))
+ #  define PRIVATE_MP_DEPRECATED_PRAGMA(s) _Pragma(#s)
+ #  define MP_DEPRECATED_PRAGMA(s) PRIVATE_MP_DEPRECATED_PRAGMA(GCC warning s)
  #elif defined(_MSC_VER) && _MSC_VER >= 1500
  #  define MP_DEPRECATED(x) __declspec(deprecated("replaced by " #x))
  #  define MP_DEPRECATED_PRAGMA(s) __pragma(message(s))
  #else
  #  define MP_DEPRECATED(s)
  #  define MP_DEPRECATED_PRAGMA(s)
  #endif
```

**Comment by minad** [comment]
@karel-m This has been fixed in https://github.com/libtom/libtommath/pull/262/commits/2a9148e25ffe709594d9bc6c66324ef80b31f61e in #262

**Comment by karel-m** [comment]
The #262 is not correct

**Comment by minad** [comment]
@karel-m What do you mean? Is not correct?

**Comment by karel-m** [comment]
`#if defined(__GNUC__) && __GNUC__ >= 4` is wrong and will cause a failure with e,g, gcc-4.4 see my research above.

**Comment by minad** [comment]
@karel-m Thank you. Fixed in #262

**Comment by karel-m** [comment]
Great, closing this one.
